modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala

```diff
@@ -44,29 +44,67 @@ package object engine {
     * `Running` status.
     */
   def switch(q: EventQueue)(st: Status): Engine[QState] =
-    modify (QState.status.set(_, st)) *> { if (st == Status.Running) next(q) else get }
+    modify(QState.status.set(_, st)) *> {
+      if (st == Status.Running) prime *> execute(q) *> get
+      else get
+    }
+
+  def prime: Engine[Unit] =
+    gets(QState.prime(_)) >>= {
+      case None => unit
+      case Some(qs) => put(qs)
+    }
 
   /**
     * Adds the `Current` `Execution` to the completed `Queue`, makes the next
     * pending `Execution` the `Current` one, and initiates the actual execution.
     *
     * If there are no more pending `Execution`s, it emits the `Finished` event.
     */
-  // XXX: Handle trying to `next` when there are actions pending in Current.
   def next(q: EventQueue): Engine[QState] =
     (gets(QState.next(_)) >>= {
-       // This should only happen when Queue is empty.
+       // No more Executions left
        case None => send(q)(finished)
-       case Some((actions, s)) => put(s) *> execute(q)(actions)
+         // Execution completed, execute next actions
+       case Some(qs) => put(qs) *> execute(q)
      }) *> get
 
   /**
+    * Checks the `Status` is `Running` and executes all actions in the `Current`
+    * `Execution` in parallel. When all are done it emits the `Executed` event.
+    * It also updates the `QState` as needed.
+    */
+  private def execute(q: EventQueue): Engine[Unit] = {
+
+    // Send the expected event when action is executed
+    def act(t: (Action, Int)): Task[Unit] = {
+      val (action, i) = t
+      action >>= {
+        case Result.OK => q.enqueueOne(completed(i))
+        case Result.Error => q.enqueueOne(failed(i))
+      }
+    }
+
+    status >>= {
+      case Status.Waiting => unit
+      case Status.Running => (
+        gets(_.current.actions) >>= (
+          actions => Nondeterminism[Task].gatherUnordered(
+            actions.zipWithIndex.map(act(_))
+          ).liftM[EngineStateT]
```

Comment:
This is just for lifting a `Task` into `Engine`: `Task[A] ~> Engine[A]`. `Engine` is a monad transformer of just 1 layer, the StateT with the Seqexec State over `Task` as the base monad.

What I got from the discussion we had a while back, with @swalker2m and @tpolecat too, is that 1 layer is manageable but more than 1 layer become too cumbersome in Scala. It seems that when the need for more transformer layers arise all the solutions lean towards the use of Free Monads for cleaner and more composable effects.

I wanted to experiment first with a simple toy example combining 2 effects using Free monads and coproducts before trying to do it in the Seqexec code, but so far I could get away with just 1 transformer layer without too much trouble.

However, I don't think the code will be able to scale in complexity properly with just 1 layer. For example, another layer I wish we had is a Reader monad (a.k.a Kleisli) with the *scalaz-stream* queue so that we don't have to be passing the queue explicitly as a parameter all over the code. Also, a *Reader* will be inevitable when/if we need to access a global configuration object from the engine.

So at some point all those `liftM` should go away.